pkg/agent/proxy/proxier_test.go

```diff
@@ -1025,213 +1203,833 @@ func testSessionAffinity(t *testing.T, svcExternalIPs net.IP, svcIP net.IP, epIP
+
+	if svcType == corev1.ServiceTypeNodePort || svcType == corev1.ServiceTypeLoadBalancer {
+		mockOFClient.EXPECT().InstallServiceFlows(groupID, vIP, uint16(svcNodePort), bindingProtocol, uint16(0), false, corev1.ServiceTypeNodePort).Times(1)
+		mockRouteClient.EXPECT().AddNodePort(nodePortAddresses, uint16(svcNodePort), bindingProtocol).Times(1)
+
+		s1 := mockOFClient.EXPECT().UninstallServiceFlows(vIP, uint16(svcNodePort), bindingProtocol)
+		mockRouteClient.EXPECT().DeleteNodePort(nodePortAddresses, uint16(svcNodePort), bindingProtocol).Times(1)
+		s2 := mockOFClient.EXPECT().InstallServiceFlows(groupID, vIP, uint16(svcNodePort+1), bindingProtocol, uint16(0), false, corev1.ServiceTypeNodePort).Times(1)
+		mockRouteClient.EXPECT().AddNodePort(nodePortAddresses, uint16(svcNodePort+1), bindingProtocol).Times(1)
+		s2.After(s1)
+	}
+	if svcType == corev1.ServiceTypeLoadBalancer {
+		mockOFClient.EXPECT().InstallServiceFlows(groupID, loadBalancerIP, uint16(svcPort), bindingProtocol, uint16(0), false, corev1.ServiceTypeLoadBalancer).Times(1)
+		mockRouteClient.EXPECT().AddLoadBalancer([]string{loadBalancerIP.String()}).Times(1)
 
+		mockOFClient.EXPECT().UninstallServiceFlows(loadBalancerIP, uint16(svcPort), bindingProtocol)
+		mockOFClient.EXPECT().InstallServiceFlows(groupID, loadBalancerIP, uint16(svcPort), bindingProtocol, uint16(0), false, corev1.ServiceTypeLoadBalancer).Times(1)
+		mockRouteClient.EXPECT().DeleteLoadBalancer([]string{loadBalancerIP.String()}).Times(1)
+		mockRouteClient.EXPECT().AddLoadBalancer([]string{loadBalancerIP.String()}).Times(1)
+	}
+
+	fp.syncProxyRules()
+	fp.serviceChanges.OnServiceUpdate(svc, updatedSvc)
 	fp.syncProxyRules()
+}
 
-	serviceNew := makeTestService(svcPortName.Namespace, svcPortName.Name, func(svc *corev1.Service) {
-		svc.Spec.ClusterIP = svcIP.String()
-		svc.Spec.Ports = []corev1.ServicePort{{
-			Name:     svcPortName.Port,
-			Port:     int32(svcPort2),
-			Protocol: corev1.ProtocolTCP,
-		}}
+func TestServiceNodePortUpdate(t *testing.T) {
+	t.Run("IPv4", func(t *testing.T) {
+		t.Run("NodePort", func(t *testing.T) {
+			testServiceNodePortUpdate(t, nodePortAddressesIPv4, svc1IPv4, nil, ep1IPv4, corev1.ServiceTypeNodePort, false)
+		})
+		t.Run("LoadBalancer", func(t *testing.T) {
+			testServiceNodePortUpdate(t, nodePortAddressesIPv4, svc1IPv4, loadBalancerIPv4, ep1IPv4, corev1.ServiceTypeLoadBalancer, false)
+		})
+	})
+	t.Run("IPv6", func(t *testing.T) {
+		t.Run("NodePort", func(t *testing.T) {
+			testServiceNodePortUpdate(t, nodePortAddressesIPv6, svc1IPv6, nil, ep1IPv6, corev1.ServiceTypeNodePort, true)
+		})
+		t.Run("LoadBalancer", func(t *testing.T) {
+			testServiceNodePortUpdate(t, nodePortAddressesIPv6, svc1IPv6, loadBalancerIPv6, ep1IPv6, corev1.ServiceTypeLoadBalancer, true)
+		})
 	})
+}
+
+func testServiceExternalTrafficPolicyUpdate(t *testing.T,
+	nodePortAddresses []net.IP,
+	svcIP,
+	loadBalancerIP,
+	ep1IP,
+	ep2IP net.IP,
```

Comment:
Better to place them in one line if they share the same type, or add type for each parameter, please double check other occurrences.

Reply:
Will do. I prefer adding type for each parameter IIRC, all parameters should be in one line or each parameter in a line.